From 9aa9a23e0693b60dec1734eef5dfdec7393e9c18 Mon Sep 17 00:00:00 2001
From: "henry@bluestacks.com" <henry@bluestacks.com>
Date: Sat, 9 Dec 2023 20:48:10 +0800
Subject: [PATCH 103/224] [A13] Making sure that bluestacks packages cannot be
 disabled by users even with other commandline options like disable-user or
 disable-until-used.

Related commitId:
Pie:75001d63d472658aa0c86551cb172c5b428ad4f7
Rvc:80b5a0af8b585c0173f566972b25d3193619242d
---
 .../com/android/server/pm/PackageManagerService.java   | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/services/core/java/com/android/server/pm/PackageManagerService.java b/services/core/java/com/android/server/pm/PackageManagerService.java
index 6bd1a3f6f218..2685bb77de6f 100644
--- a/services/core/java/com/android/server/pm/PackageManagerService.java
+++ b/services/core/java/com/android/server/pm/PackageManagerService.java
@@ -3759,6 +3759,7 @@ public class PackageManagerService implements PackageSender, TestUtilityService
         final int targetSize = settings.size();
         for (int i = 0; i < targetSize; i++) {
             final int newState = settings.get(i).getEnabledState();
+            String packageName = settings.get(i).getPackageName();
             if (!(newState == COMPONENT_ENABLED_STATE_DEFAULT
                     || newState == COMPONENT_ENABLED_STATE_ENABLED
                     || newState == COMPONENT_ENABLED_STATE_DISABLED
@@ -3766,6 +3767,15 @@ public class PackageManagerService implements PackageSender, TestUtilityService
                     || newState == COMPONENT_ENABLED_STATE_DISABLED_UNTIL_USED)) {
                 throw new IllegalArgumentException("Invalid new component state: " + newState);
             }
+            // If user is trying to disable one of our packages, don't allow it.
+            if (packageName.startsWith("com.bluestacks") || "com.uncube.account".equalsIgnoreCase(packageName)) {
+                if (newState == COMPONENT_ENABLED_STATE_DISABLED
+                    || newState == COMPONENT_ENABLED_STATE_DISABLED_USER
+                    || newState == COMPONENT_ENABLED_STATE_DISABLED_UNTIL_USED) {
+                    Slog.d(TAG, "trying to disable bluestacks package, permission denied - returning");
+                    return;
+                }
+            }
         }
         if (targetSize > 1) {
             final ArraySet<String> checkDuplicatedPackage = new ArraySet<>();
